### PR TITLE
save args to txt file

### LINKED ISSUE #214: Print arguments in a seperate file
Currently we log the arguments used (the lines right after `------------------------ arguments ------------------------`) in stdout that we catch inside `main_log.txt`.

I think we should also store them in a seperate file to check the config seperatly.

### PR DESCRIPTION
**Pull request by bhavitvyamalik**
I decided to work on it after there was inactivity on #214

I wasn't sure where to save the file, so I saved it in the current directory only. Since we have to use it to check the config separately, I thought of naming the file with current timestamp. I hope that's okay. Please let me know if any other changes are required for this!

### CODE REVIEW

**Review by thomasw21 — COMMENTED** [review]
Hey! Thanks for taking a look! I think it makes sense to store them in something we'd call `log_path` or something. We usually store all the logs like that: https://github.com/bigscience-workshop/bigscience/blob/9920b0567d241fc7625d79f0cb134c0689ccdb28/train/tr3-1B3-baseline/tr3o-1B3-pile-no-dropout.slurm#L193 . I think it might be nice to introduce that log_path as an argument to the script, and then save the args in a file at that location.

Otherwise this looks good to me

File: megatron/arguments.py [hunk 1]
```diff
@@ -328,8 +329,11 @@ def _print_args(args):
         for arg in vars(args):
             dots = '.' * (48 - len(arg))
             str_list.append('  {} {} {}'.format(arg, dots, getattr(args, arg)))
-        for arg in sorted(str_list, key=lambda x: x.lower()):
-            print(arg, flush=True)
+
+        with open(f'args_{time.strftime("%Y%m%d-%H%M%S")}.txt', 'w') as f:
+            for arg in sorted(str_list, key=lambda x: x.lower()):
+                f.write(str(arg)+"\n")
```
[the remark this answers is not in the document]
Comment (bhavitvyamalik) [reply megatron/arguments.py:335]: just to be on the safe side I did this. No issues, I'll remove it

**Comment by bhavitvyamalik** [comment]
Pushed the changes suggested by you. Also, are there any issues/current ongoing work here I can contribute to? I want to learn and contribute to how you guys are training models at scale.

File: megatron/arguments.py [hunk 2]
```diff
@@ -328,8 +329,11 @@ def _print_args(args):
         for arg in vars(args):
             dots = '.' * (48 - len(arg))
             str_list.append('  {} {} {}'.format(arg, dots, getattr(args, arg)))
-        for arg in sorted(str_list, key=lambda x: x.lower()):
-            print(arg, flush=True)
+
+        with open(os.path.join(args.log_path,f'args_{time.strftime("%Y-%m-%dT%H:%M:%S")}.txt'), 'w') as f:
```
Comment (thomasw21) [inline megatron/arguments.py:333]:
```python
>>> import os
>>> os.path.join(None, "file.txt")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.9/posixpath.py", line 76, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Can you please add a test? The test should probably be in https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/tests/test_training.py

**Review by thomasw21 — COMMENTED** [review]
Please add a test.

Concerning new issues, I can't think of urgent things of the top of my head. I think you can look at issues that are currently here. Also if you spot any weird things, feel free to open an issue and fix them.

File: megatron/arguments.py (hunk 2 above)
Comment (bhavitvyamalik) [reply megatron/arguments.py:333]: I should pass `log-path` to [`args`](https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/4ed19efb2c50ecc35f693360e2488a5cf3e4ca6d/tests/test_training.py#L348) and then `assert` it's not `None`?

Comment (bhavitvyamalik) [reply megatron/arguments.py:333]: Basically make it an optional parameter? If given, then it'll save `txt` file, else skip?

Comment (thomasw21) [reply megatron/arguments.py:333]: Yeah that also works

Comment (thomasw21) [reply megatron/arguments.py:333]: That's breaking backward compatibility. Essentially you're suggesting we crash runs that don't specify that line no?

Comment (bhavitvyamalik) [reply megatron/arguments.py:333]: My bad! I mean let's have `if` before dumping args in the text file. If `log-path` was passed, then it'll save txt file, else skip, thereby ensuring backward compatibility. Inside `test_training.py`, in `args` variable I can set `log-path` to both `None` and `str`(for different `args`) in order to check if runs work in both cases

Comment (thomasw21) [reply megatron/arguments.py:333]: sounds good

File: tests/test_training.py [hunk 3]
```diff
@@ -304,6 +304,7 @@ def test_training_prefix_lm_all(self, loss_on_targets_only, reweight_loss_based_
         src_dir = self.src_dir
         data_dir = f"{self.data_dir}/gpt2"
         output_dir = self.get_auto_remove_tmp_dir() # "./xxx", after=False)
+        logs_dir = f"{output_dir}/logs"
```
Comment (bhavitvyamalik) [inline tests/test_training.py:307]: I've added `logs` directory in `output_dir`

**Review by thomasw21 — APPROVED** [review]
Aweseome!

**Comment by stas00** [comment]
this PR introduced broken tests. https://github.com/bigscience-workshop/Megatron-DeepSpeed/runs/4824420122?check_suite_focus=true fixed in https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/232

I know we have an issue with CI not being able to run on forked PRs due to AWS, so please try to branch on the main repo instead to overcome this issue. Please let me know if you need write access to be able to do that.

If this didn't happen and the CI failed to start please run the tests locally and only merge on success.

Thank you.

**Comment by bhavitvyamalik** [comment]
Sure, I'll keep that in mind next time onwards. I thought there must be issue with the CI hence didn't cross check my code. Before my next PR I'll let you know if write access is required. Thank you, @stas00!

**Comment by thomasw21** [comment]
My bad, I had launched the tests but they failed due to AWS credentials setup (https://github.com/bigscience-workshop/Megatron-DeepSpeed/runs/4797062928?check_suite_focus=true) so I just assumed it would work out of the box.

**Comment by stas00** [comment]
It's all good, guys! We are just trying to find the easiest way to make things work for all.

Is there a way to tell github not to allow PRs from non-main repo branches? Perhaps if there is and we enable that and give those who need commit access then this issue will be solved.

the only other solution is setting up a 24/7 runner which would be much more expensive and need to invest a lot of time to set it up again.

**Pull request merged**